File: Arrays/P023.py
def brute(nums):
    n=len(nums)
    max_sum=float('-inf')
    for i in range(n):
        suma=0
        for j in range(i,n):
            suma+=nums[j]
            max_sum=max(max_sum,suma)
    return max_sum

def optimal(nums):
    n=len(nums)
    cur_sum=max_sum=nums[0]
    for i in range(1,n):
        cur_sum=max(cur_sum+nums[i],nums[i])
        max_sum=max(max_sum,cur_sum)
    return max_sum

if __name__=='__main__':
    nums=input("Enter array: ").split()
    nums=list(map(int,nums))
    print("Brute: ",brute(nums))
    print("Optimal: ",optimal(nums))

File: Striver_C++_Problems/Arrays/P023.cpp
// P023.cpp - Maximum subarray sum

#include <bits/stdc++.h>
using namespace std;

long long brute(const vector<int>& nums) {
    int n = nums.size();
    long long max_sum = LLONG_MIN;
    for (int i = 0; i < n; i++) {
        long long suma = 0;
        for (int j = i; j < n; j++) {
            suma += nums[j];
            max_sum = max(max_sum, suma);
        }
    }
    return max_sum;
}

long long optimal(const vector<int>& nums) {
    int n = nums.size();
    long long cur_sum = nums[0];
    long long max_sum = nums[0];
    for (int i = 1; i < n; i++) {
        cur_sum = max(cur_sum + nums[i], (long long)nums[i]);
        max_sum = max(max_sum, cur_sum);
    }
    return max_sum;
}

int main() {
    cout << "Enter array: ";
    string line;
    getline(cin, line);
    stringstream ss(line);
    vector<int> nums;
    int x;
    while (ss >> x) nums.push_back(x);

    cout << "Brute:  " << brute(nums) << "\n";
    cout << "Optimal: " << optimal(nums) << "\n";
    return 0;
}
